Fix deletion of non-uploaded languages

orangejenny opened the pull request:
PR into https://github.com/dimagi/commcare-hq/pull/23658

Fixes a bug where when a single-language file is uploaded, all translations in *other* languages are deleted.

(product team, ignore this)

orangejenny committed:
Renamed upload_no_change_data and removed duplicate headers list

orangejenny committed:
Added failing test

orangejenny committed:
Fixed bug

orangejenny committed:
Added failed text for node deletion bug

orangejenny committed:
Fixed bug with deleting nodes when only uploading single sheet

orangejenny committed:
Removed fallback logic if there's only a single language

orangejenny committed:
Moved condition; committed too soon

orangejenny commented:
Added another few commits to fix a bug where nodes were being deleted from forms when a single-language upload didn't provide a translation. Nodes should only be deleted if the upload is providing all languages and there are no values for any of them.

mkangia commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -158,13 +158,13 @@ def _add_or_remove_translations(self, lang, row):
         text_node = self._get_text_node(translation_node, label_id)
         for trans_type, new_translation in translations.items():
-            if not new_translation:
+            if not new_translation and len(self.langs) > 1:
```
If there is actually just one lang in the app this block would never run?

orangejenny replied:
Yes - if there's one language and the translation is blank, there aren't any other languages to fall back to, so this block is irrelevant.

mkangia commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -158,13 +158,13 @@ def _add_or_remove_translations(self, lang, row):
         translations = self._get_translations_for_row(row, lang)
         translation_node = self.itext.find("./{f}translation[@lang='%s']" % lang)
-        keep_value_node = any(v for k, v in translations.items())
+        keep_value_node = any(v for k, v in translations.items()) or set(self.langs) != set(self.app.langs)
```
i think we should put this in a variable with a nice name like
`all_app_langs_in_upload = set(self.langs) != set(self.app.langs)` 

Also better to keep this before the translations iterations?

orangejenny replied:
I was thinking about setting a flag at the class level to indicate whether this is single-sheet or multi-sheet - adding in 31aecb9c9314389055242c32512379ef5d196b9e. Doing that makes this clearer, I think.

Also swapped the order of the conditions.

orangejenny committed:
Added is_multi_sheet flag

orangejenny committed:
Fixed test data.

In the multi-sheet upload, there's a value for item2, so there should be a value
for item2 in the xml in both english (provided) and french (fallback).

orangejenny committed:
Merge branch 'jls/upload-mm-translations' into jls/upload-mm-translations-fix-deletions
